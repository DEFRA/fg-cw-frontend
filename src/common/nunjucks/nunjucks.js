import Vision from "@hapi/vision";
import Nunjucks from "nunjucks";

import { glob } from "node:fs/promises";
import { config } from "../config.js";
import { filters } from "../format/format-filters.js";
import { context } from "./context/context.js";
import { formatDate } from "./filters/format-date.js";

const njkPaths = [
  "node_modules/govuk-frontend/dist/",
  ...(await Array.fromAsync(
    glob(["**/views", "**/*(layouts|components|pages|partials)/"], {
      exclude: ["node_modules", "*.*"],
    }),
  )),
];

const viewPaths = await Array.fromAsync(
  glob(["**/views"], {
    exclude: ["node_modules", "*.*"],
  }),
);

export const environment = Nunjucks.configure(njkPaths, {
  autoescape: true,
  throwOnUndefined: false,
  trimBlocks: true,
  lstripBlocks: true,
  watch: config.get("nunjucks.watch"),
  noCache: config.get("nunjucks.noCache"),
  filter: {
    formatDate,
  },
});

environment.addFilter("formatDate", formatDate);
environment.addFilter("fixed", filters.fixed);

export const nunjucks = {
  plugin: Vision,
  options: {
    engines: {
      njk: {
        compile(src, options) {
          const template = Nunjucks.compile(src, options.environment);
          return (ctx) => template.render(ctx);
        },
      },
    },
    compileOptions: {
      environment,
    },
    relativeTo: config.get("root"),
    path: viewPaths,
    isCached: config.get("isProduction"),
    context,
  },
};
