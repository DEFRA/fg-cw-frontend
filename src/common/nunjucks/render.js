import camelCase from "lodash/camelCase.js";
import { environment } from "./nunjucks.js";

/**
 * @param {string} name
 * @param {object} params
 * @param {string} [block]
 */
export const render = (name, params, block) => {
  const macro = {
    path: `${name}/macro.njk`,
    name: camelCase(name.split("-")),
    params: JSON.stringify(params, null, 2),
  };

  const callComponent = block
    ? `{%- call ${macro.name}(${macro.params}) -%}${block}{%- endcall -%}`
    : `{{- ${macro.name}(${macro.params}) -}}`;

  return environment.renderString(
    `{%- from "${macro.path}" import ${macro.name} -%}
    ${callComponent}
    `,
    {},
  );
};
