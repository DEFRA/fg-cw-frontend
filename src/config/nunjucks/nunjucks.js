import { fileURLToPath } from 'node:url'
import path from 'path'
import nunjucks from 'nunjucks'
import hapiVision from '@hapi/vision'

import { config } from '../config.js'
import { context } from './context/context.js'
import * as filters from './filters/filters.js'
import * as globals from './globals.js'

const dirname = path.dirname(fileURLToPath(import.meta.url))
const nunjucksEnvironment = nunjucks.configure(
  [
    'node_modules/govuk-frontend/dist/',
    path.resolve(dirname, '../../server/common/templates'),
    path.resolve(dirname, '../../server/common/components'),
    path.resolve(dirname, '../../case-management/views'),
    path.resolve(dirname, '../../case-management/views/layouts'),
    path.resolve(dirname, '../../case-management/views/components'),
    path.resolve(dirname, '../../case-management/views/pages'),
    path.resolve(dirname, '../../case-management/views/partials'),
    path.resolve(dirname, '../../common/presentation')
  ],
  {
    autoescape: true,
    throwOnUndefined: false,
    trimBlocks: true,
    lstripBlocks: true,
    watch: config.get('nunjucks.watch'),
    noCache: config.get('nunjucks.noCache')
  }
)

export const nunjucksConfig = {
  plugin: hapiVision,
  options: {
    engines: {
      njk: {
        compile(src, options) {
          const template = nunjucks.compile(src, options.environment)
          return (ctx) => template.render(ctx)
        }
      }
    },
    compileOptions: {
      environment: nunjucksEnvironment
    },
    relativeTo: path.resolve(dirname, '../..'),
    path: ['server', 'case-management/views', 'common/presentation'],
    isCached: config.get('isProduction'),
    context
  }
}

Object.entries(globals).forEach(([name, global]) => {
  nunjucksEnvironment.addGlobal(name, global)
})

Object.entries(filters).forEach(([name, filter]) => {
  nunjucksEnvironment.addFilter(name, filter)
})
