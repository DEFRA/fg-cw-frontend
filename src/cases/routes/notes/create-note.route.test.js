import hapi from "@hapi/hapi";
import { load } from "cheerio";
import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import { createMockCaseData } from "../../../../test/data/case-test-data.js";
import { nunjucks } from "../../../common/nunjucks/nunjucks.js";
import { addNoteToCaseUseCase } from "../../use-cases/add-note-to-case.use-case.js";
import { findCaseByIdUseCase } from "../../use-cases/find-case-by-id.use-case.js";
import { createNoteRoute } from "./create-note.route.js";

vi.mock("../../use-cases/find-case-by-id.use-case.js");
vi.mock("../../use-cases/add-note-to-case.use-case.js");

describe("createNoteRoute", () => {
  let server;

  beforeAll(async () => {
    server = hapi.server();
    server.route(createNoteRoute);
    await server.register([nunjucks]);

    await server.initialize();
  });

  afterAll(async () => {
    await server.stop();
  });

  it("calls use case and redirects on successful submission", async () => {
    addNoteToCaseUseCase.mockResolvedValue();

    const { statusCode, headers } = await server.inject({
      method: "POST",
      url: "/cases/68495db5afe2d27b09b2ee47/notes",
      payload: {
        text: "This is a valid note",
      },
    });

    expect(addNoteToCaseUseCase).toHaveBeenCalledWith({
      caseId: "68495db5afe2d27b09b2ee47",
      text: "This is a valid note",
    });

    expect(statusCode).toEqual(302);
    expect(headers.location).toEqual("/cases/68495db5afe2d27b09b2ee47/notes");
  });

  it("shows validation error when text is empty", async () => {
    const mockCaseData = createMockCaseData();
    findCaseByIdUseCase.mockResolvedValue(mockCaseData);

    const { statusCode, result } = await server.inject({
      method: "POST",
      url: "/cases/68495db5afe2d27b09b2ee47/notes",
      payload: {
        text: "",
      },
    });

    expect(findCaseByIdUseCase).toHaveBeenCalledWith(
      "68495db5afe2d27b09b2ee47",
    );
    expect(addNoteToCaseUseCase).not.toHaveBeenCalled();
    expect(statusCode).toEqual(200);

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });

  it("shows validation error when text is only whitespace", async () => {
    const mockCaseData = createMockCaseData();
    findCaseByIdUseCase.mockResolvedValue(mockCaseData);

    const { statusCode, result } = await server.inject({
      method: "POST",
      url: "/cases/68495db5afe2d27b09b2ee47/notes",
      payload: {
        text: "   \n\t  ",
      },
    });

    expect(findCaseByIdUseCase).toHaveBeenCalledWith(
      "68495db5afe2d27b09b2ee47",
    );
    expect(addNoteToCaseUseCase).not.toHaveBeenCalled();
    expect(statusCode).toEqual(200);

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });

  it("shows validation error when text is missing", async () => {
    const mockCaseData = createMockCaseData();
    findCaseByIdUseCase.mockResolvedValue(mockCaseData);

    const { statusCode, result } = await server.inject({
      method: "POST",
      url: "/cases/68495db5afe2d27b09b2ee47/notes",
      payload: {
        // text is missing
      },
    });

    expect(findCaseByIdUseCase).toHaveBeenCalledWith(
      "68495db5afe2d27b09b2ee47",
    );
    expect(addNoteToCaseUseCase).not.toHaveBeenCalled();
    expect(statusCode).toEqual(200);

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });

  it("shows save error when save fails", async () => {
    const mockCaseData = createMockCaseData();
    findCaseByIdUseCase.mockResolvedValue(mockCaseData);
    addNoteToCaseUseCase.mockRejectedValue(new Error("API Error"));

    const { statusCode, result } = await server.inject({
      method: "POST",
      url: "/cases/68495db5afe2d27b09b2ee47/notes",
      payload: {
        text: "This will fail to save",
      },
    });

    expect(addNoteToCaseUseCase).toHaveBeenCalledWith({
      caseId: "68495db5afe2d27b09b2ee47",
      text: "This will fail to save",
    });
    expect(findCaseByIdUseCase).toHaveBeenCalledWith(
      "68495db5afe2d27b09b2ee47",
    );
    expect(statusCode).toEqual(200);

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });
});
