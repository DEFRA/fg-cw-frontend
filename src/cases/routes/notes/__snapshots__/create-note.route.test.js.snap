// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createNoteRoute > shows save error when save fails 1`] = `
"
  
<div class="govuk-!-margin-bottom-6 app-case-banner">
  <div class="govuk-width-container">
    <a href="/cases" class="govuk-back-link">Back to cases</a>

    <h1 class="govuk-heading-m">
      Valley Farm
    </h1>

    <div>
      <div class="app-case-row">
      </div>
    </div>
  </div>
</div>


    <div class="govuk-error-summary" data-module="govuk-error-summary">
  <div role="alert">
    <h2 class="govuk-error-summary__title">
      There is a problem
    </h2>
    <div class="govuk-error-summary__body">
        <ul class="govuk-list govuk-error-summary__list">
          <li>
            There was a problem saving the note. Please try again.
          </li>
        </ul>
    </div>
  </div>
</div>


  <div class="govuk-body">
    
  <div class="govuk-service-navigation" data-module="govuk-service-navigation">
      <div class="govuk-width-container">

    <div class="govuk-service-navigation__container">
      

      
        <nav aria-label="Menu" class="govuk-service-navigation__wrapper">
          <button type="button" class="govuk-service-navigation__toggle govuk-js-service-navigation-toggle" aria-controls="navigation" hidden="">
            Menu
          </button>

          <ul class="govuk-service-navigation__list" id="navigation">

            
              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47">
                                    
Tasks
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/case-details">
                                    
Case Details
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item govuk-service-navigation__item--active">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/notes" aria-current="true">
                                    
                  <strong class="govuk-service-navigation__active-fallback">Notes</strong>

                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/timeline">
                                    
Timeline
                  </a>
              </li>

            </ul>
        </nav>
    </div>

    </div>

  </div>



    <form action="/cases/68495db5afe2d27b09b2ee47/notes" method="post">
      <input type="hidden" name="type" value="NOTE_ADDED">

      <div class="govuk-form-group">
  <label class="govuk-label govuk-label govuk-label--l govuk-!-margin-top-6" for="text">
    Add a note
  </label>
  <div id="text-hint" class="govuk-hint">
    All notes will be saved for auditing purposes
  </div>
  <textarea class="govuk-textarea" id="text" name="text" rows="5" aria-describedby="text-hint">This will fail to save</textarea>
</div>


      <button type="submit" class="govuk-button" data-module="govuk-button">
  Save
</button>
    </form>
  </div>
        "
`;

exports[`createNoteRoute > shows validation error when text is empty 1`] = `
"
  
<div class="govuk-!-margin-bottom-6 app-case-banner">
  <div class="govuk-width-container">
    <a href="/cases" class="govuk-back-link">Back to cases</a>

    <h1 class="govuk-heading-m">
      Valley Farm
    </h1>

    <div>
      <div class="app-case-row">
      </div>
    </div>
  </div>
</div>


    <div class="govuk-error-summary" data-module="govuk-error-summary">
  <div role="alert">
    <h2 class="govuk-error-summary__title">
      There is a problem
    </h2>
    <div class="govuk-error-summary__body">
        <ul class="govuk-list govuk-error-summary__list">
          <li>
            <a href="#text">You must enter a note</a>
          </li>
        </ul>
    </div>
  </div>
</div>


  <div class="govuk-body">
    
  <div class="govuk-service-navigation" data-module="govuk-service-navigation">
      <div class="govuk-width-container">

    <div class="govuk-service-navigation__container">
      

      
        <nav aria-label="Menu" class="govuk-service-navigation__wrapper">
          <button type="button" class="govuk-service-navigation__toggle govuk-js-service-navigation-toggle" aria-controls="navigation" hidden="">
            Menu
          </button>

          <ul class="govuk-service-navigation__list" id="navigation">

            
              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47">
                                    
Tasks
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/case-details">
                                    
Case Details
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item govuk-service-navigation__item--active">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/notes" aria-current="true">
                                    
                  <strong class="govuk-service-navigation__active-fallback">Notes</strong>

                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/timeline">
                                    
Timeline
                  </a>
              </li>

            </ul>
        </nav>
    </div>

    </div>

  </div>



    <form action="/cases/68495db5afe2d27b09b2ee47/notes" method="post">
      <input type="hidden" name="type" value="NOTE_ADDED">

      <div class="govuk-form-group govuk-form-group--error">
  <label class="govuk-label govuk-label govuk-label--l govuk-!-margin-top-6" for="text">
    Add a note
  </label>
  <div id="text-hint" class="govuk-hint">
    All notes will be saved for auditing purposes
  </div>
  <p id="text-error" class="govuk-error-message">
    <span class="govuk-visually-hidden">Error:</span> You must enter a note
  </p>
  <textarea class="govuk-textarea govuk-textarea--error" id="text" name="text" rows="5" aria-describedby="text-hint text-error"></textarea>
</div>


      <button type="submit" class="govuk-button" data-module="govuk-button">
  Save
</button>
    </form>
  </div>
        "
`;

exports[`createNoteRoute > shows validation error when text is missing 1`] = `
"
  
<div class="govuk-!-margin-bottom-6 app-case-banner">
  <div class="govuk-width-container">
    <a href="/cases" class="govuk-back-link">Back to cases</a>

    <h1 class="govuk-heading-m">
      Valley Farm
    </h1>

    <div>
      <div class="app-case-row">
      </div>
    </div>
  </div>
</div>


    <div class="govuk-error-summary" data-module="govuk-error-summary">
  <div role="alert">
    <h2 class="govuk-error-summary__title">
      There is a problem
    </h2>
    <div class="govuk-error-summary__body">
        <ul class="govuk-list govuk-error-summary__list">
          <li>
            <a href="#text">You must enter a note</a>
          </li>
        </ul>
    </div>
  </div>
</div>


  <div class="govuk-body">
    
  <div class="govuk-service-navigation" data-module="govuk-service-navigation">
      <div class="govuk-width-container">

    <div class="govuk-service-navigation__container">
      

      
        <nav aria-label="Menu" class="govuk-service-navigation__wrapper">
          <button type="button" class="govuk-service-navigation__toggle govuk-js-service-navigation-toggle" aria-controls="navigation" hidden="">
            Menu
          </button>

          <ul class="govuk-service-navigation__list" id="navigation">

            
              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47">
                                    
Tasks
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/case-details">
                                    
Case Details
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item govuk-service-navigation__item--active">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/notes" aria-current="true">
                                    
                  <strong class="govuk-service-navigation__active-fallback">Notes</strong>

                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/timeline">
                                    
Timeline
                  </a>
              </li>

            </ul>
        </nav>
    </div>

    </div>

  </div>



    <form action="/cases/68495db5afe2d27b09b2ee47/notes" method="post">
      <input type="hidden" name="type" value="NOTE_ADDED">

      <div class="govuk-form-group govuk-form-group--error">
  <label class="govuk-label govuk-label govuk-label--l govuk-!-margin-top-6" for="text">
    Add a note
  </label>
  <div id="text-hint" class="govuk-hint">
    All notes will be saved for auditing purposes
  </div>
  <p id="text-error" class="govuk-error-message">
    <span class="govuk-visually-hidden">Error:</span> You must enter a note
  </p>
  <textarea class="govuk-textarea govuk-textarea--error" id="text" name="text" rows="5" aria-describedby="text-hint text-error"></textarea>
</div>


      <button type="submit" class="govuk-button" data-module="govuk-button">
  Save
</button>
    </form>
  </div>
        "
`;

exports[`createNoteRoute > shows validation error when text is only whitespace 1`] = `
"
  
<div class="govuk-!-margin-bottom-6 app-case-banner">
  <div class="govuk-width-container">
    <a href="/cases" class="govuk-back-link">Back to cases</a>

    <h1 class="govuk-heading-m">
      Valley Farm
    </h1>

    <div>
      <div class="app-case-row">
      </div>
    </div>
  </div>
</div>


    <div class="govuk-error-summary" data-module="govuk-error-summary">
  <div role="alert">
    <h2 class="govuk-error-summary__title">
      There is a problem
    </h2>
    <div class="govuk-error-summary__body">
        <ul class="govuk-list govuk-error-summary__list">
          <li>
            <a href="#text">You must enter a note</a>
          </li>
        </ul>
    </div>
  </div>
</div>


  <div class="govuk-body">
    
  <div class="govuk-service-navigation" data-module="govuk-service-navigation">
      <div class="govuk-width-container">

    <div class="govuk-service-navigation__container">
      

      
        <nav aria-label="Menu" class="govuk-service-navigation__wrapper">
          <button type="button" class="govuk-service-navigation__toggle govuk-js-service-navigation-toggle" aria-controls="navigation" hidden="">
            Menu
          </button>

          <ul class="govuk-service-navigation__list" id="navigation">

            
              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47">
                                    
Tasks
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/case-details">
                                    
Case Details
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item govuk-service-navigation__item--active">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/notes" aria-current="true">
                                    
                  <strong class="govuk-service-navigation__active-fallback">Notes</strong>

                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/timeline">
                                    
Timeline
                  </a>
              </li>

            </ul>
        </nav>
    </div>

    </div>

  </div>



    <form action="/cases/68495db5afe2d27b09b2ee47/notes" method="post">
      <input type="hidden" name="type" value="NOTE_ADDED">

      <div class="govuk-form-group govuk-form-group--error">
  <label class="govuk-label govuk-label govuk-label--l govuk-!-margin-top-6" for="text">
    Add a note
  </label>
  <div id="text-hint" class="govuk-hint">
    All notes will be saved for auditing purposes
  </div>
  <p id="text-error" class="govuk-error-message">
    <span class="govuk-visually-hidden">Error:</span> You must enter a note
  </p>
  <textarea class="govuk-textarea govuk-textarea--error" id="text" name="text" rows="5" aria-describedby="text-hint text-error">   
	  </textarea>
</div>


      <button type="submit" class="govuk-button" data-module="govuk-button">
  Save
</button>
    </form>
  </div>
        "
`;
